src/ray/raylet/scheduling_resources.cc

```diff
@@ -76,7 +76,11 @@ ResourceSet::ResourceSet() {}
 
 ResourceSet::ResourceSet(
     const std::unordered_map<std::string, FractionalResourceQuantity> &resource_map)
-    : resource_capacity_(resource_map) {}
+    : resource_capacity_(resource_map) {
+  for (auto const &resource_pair : resource_map) {
+    RAY_CHECK(resource_pair.second > 0);
```

Comment:
This prevents the ability to create a resource set with any zero capacity resource. One of the use cases we supported in the past is a local laptop with zero CPU allocation to prevent work assignment/force all work assignment to the remote cluster.

Reply:
so, this change would break backward compatibility

Reply:
That's actually still supported. This PR shouldn't have any API implications. The resources with quantity 0 are filtered out before getting to this constructor.

Reply:
I'm just not comfortable with a constructor that breaks the whole raylet if you pass it a fairly reasonable zero resource capacity attribute. This restricts the programmatic expressivity of the ResourceSet. This should also be extensively tested with dynamic custom resources (ESCHER), while allowing some custom resource to drop down to zero capacity. There's a possibility that somewhere in the code we make a copy of the ResourceSet that's dynamically updated to \vec{r} = [..., 0, ...] and trigger this check. I.e., I'm not as worried about the static resource case as I am about the dynamic resource case.

Reply:
We maintain the invariant that all quantities in a `ResourceSet` are positive. That means that there is a unique internal representation for a given set of resource capacities and reduces the number of cases we have to check. We implemented those semantics in https://github.com/ray-project/ray/pull/4555.

This change makes sure to enforce those semantics more clearly (and the absence of this change allowed the bug to creep in). Note that the constructor below has the same check.

Reply:
I think as long as other methods of ResourceSet that update capacity are correctly implemented, dynamic custom resources should be fine even when capacity drops to zero. As @robertnishihara pointed out, any illegal updates must be filtered out before this code is invoked. If they're not, raising a check fail here seems helpful for isolating bugs.

Reply:
@robertnishihara , I see, I suppose it makes sense to be consistent with the second constructor, which is already performing this check. Longer term, not having the ability to represent a zero in a system seems broken to me. There was this major breakthrough in mathematics when people invented zero...